lib/ex_unit/lib/ex_unit/runner.ex

```diff
@@ -107,28 +118,39 @@ defmodule ExUnit.Runner do
   end
 
   defp run_module(config, module) do
-    test_module = module.__ex_unit__()
-    EM.module_started(config.manager, test_module)
-
-    # Prepare tests, selecting which ones should
-    # run and which ones were skipped.
-    tests = prepare_tests(config, test_module.tests)
-
-    {test_module, pending, finished_tests} =
-      if Enum.all?(tests, & &1.state) do
-        # The pending tests here aren't actually run, so they're already
-        # "finished"
-        {test_module, tests, tests}
-      else
-        spawn_module(config, test_module, tests)
-      end
+    if max_failures_reached?(config.stats_pid, config.max_failures) do
+      # This is reached when spawn_modules launches several test_modules,
+      # and one is invalid, and max_failures is reached
+      send(config.runner_pid, {self(), :max_failures_reached})
+    else
+      test_module = module.__ex_unit__()
+      EM.module_started(config.manager, test_module)
+
+      # Prepare tests, selecting which ones should be run or skipped
+      tests = prepare_tests(config, test_module.tests)
+
+      {test_module, test_results} =
+        if Enum.all?(tests, & &1.state) do
+          # The pending tests here aren't actually run,
+          # so they're marked as "finished".
+          {test_module, %{pending: tests, finished: tests}}
+        else
+          spawn_module(config, test_module, tests)
+        end
 
-    # Run the pending tests. We don't actually spawn those
-    # tests but we do send the notifications to formatter.
-    Enum.each(pending, &run_test(config, &1, []))
-    test_module = %{test_module | tests: finished_tests}
-    EM.module_finished(config.manager, test_module)
-    send(config.runner_pid, {self(), :module_finished, test_module})
+      # Do not run pending tests.
+      # Just send the notifications to the formatter.
+      Enum.each(test_results.pending, fn test ->
+        EM.test_started(config.manager, test)
+        EM.test_finished(config.manager, test)
+      end)
+
+      test_module = %{test_module | tests: test_results.finished}
+      EM.module_finished(config.manager, test_module)
+
+      # message loop/2
+      send(config.runner_pid, {self(), :module_finished})
```

Comment:
I see you simplified the message by not including the module, could this be further simplified by just sending to `loop/3` the messages `:module_finished` and `:max_failures_reached`?

Btw, the comment should say `loop/3` :) But, is it necessary?

Reply:
Please let's continue sending the three element tuple. If the previous element was not necessary, it should be removed in another PR.